### DISCUSSION TITLE
Spirng boot+mybatis 批量写入慢

### DISCUSSION THREAD

**Discussion by a3225167** [apache/doris#13888] [A - General / Q&A] [answered]
最早用的是phoneix+hbase，写入速度性能还是挺不错的。
现在换成的doris，数据的批量写入比较慢。
有什么优化的方案吗？ 是否不能用mybatis 需要使用原始的jdbc呢？

**Comment by zy-kkk** [comment, marked as answer]
Spirng boot+mybatis方式以及原始JDBC方式都不推荐，相当于大量的insert into放到doris进行，这种效率是极低的，而且会引起-235问题，我们建议采用stream load方式来进行对doris的写入，而Spirng boot+mybatis可以对Doris进行查询，以及少量的改和删
stream load demo：https://github.com/apache/doris/blob/master/samples/stream_load/java/DorisStreamLoad.java

**Reply by flyzyh** [reply]
但是…………stream load不好用啊

**Reply by a3225167** [reply]
你现在用的哪个版本doris？

**Reply by flyzyh** [reply]
最新版吧，目前刚想考虑把一些需要快速查询和统计的数据入库doris。就发现stream入库真的不好用啊，很多东西想做也做不到，还是java方便。
就是太慢了。

**Reply by flyzyh** [reply]
我们的情况有点特殊，主要是把csv数据入库。
csv文件吧，文件名里有一部分也是一个参数，我们想变成一个字段入库，甚至想把这个参数做一个映射成为code，也是新的一列。
csv文件里面还有一些换行之类的，这种换行是符合csv规范的，换行在双引号之间，算是一个字符串的内容。java的CSVFormatter就能正确读取。但是hive和doris似乎就都处理不了。他们都是一行行读取，发现换行就算新一条记录。
这些情况导致我们的使用场景光靠streamload，甚至seatunnel我们也试过，非常难以实现。只能靠写程序。
程序的问题就是慢。目前这个“慢”还算可以忍受。

**Reply by sollhui** [reply]
@flyzyh 
"这种换行是符合csv规范的，换行在双引号之间，算是一个字符串的内容"这里配置一下包围符enclose:"就行了，参考https://doris.apache.org/zh-CN/docs/dev/data-operate/import/stream-load-manual/#%E5%AF%BC%E5%85%A5%E9%85%8D%E7%BD%AE%E5%8F%82%E6%95%B0

**Reply by flyzyh** [reply]
感谢，这个csv换行的问题因为之前hive入库也遇到过，写过一个把换行都干掉的小工具，所以算是绕过去了。
但是现在streamLoad又遇到了新的问题，就是header里的columns这一项该如何描述 表中的某一列（比如column_u）是由源文件中几列（比如A，B，C）拼接而来的？
是column_u=CONCAT（A,B,C）？
还是A+B+C？
拼接的时候要是再有个下划线咋算/
这个地方又卡了两天了，doris的官方文档一句带过，网上的资料也少，根本没有对应的例子。
使用curl调试的时候，真是加倍困难

**Reply by sollhui** [reply]
> 但是现在streamLoad又遇到了新的问题，就是header里的columns这一项该如何描述 表中的某一列（比如column_u）是由源文件中几列（比如A，B，C）拼接而来的？
是column_u=CONCAT（A,B,C）？
还是A+B+C？

这里是column_u=CONCAT（A,B,C）


> 拼接的时候要是再有个下划线咋算/

这个地方没看懂，能不能举个例子呢？

**Reply by liaoxin01** [reply]
MySQL [test]> desc tbl;
+-------+------+------+-------+---------+-------+
| Field | Type | Null | Key   | Default | Extra |
+-------+------+------+-------+---------+-------+
| c1    | INT  | Yes  | true  | NULL    |       |
| c2    | INT  | Yes  | false | NULL    | NONE  |
| c3    | TEXT | Yes  | false | NULL    | NONE  |
+-------+------+------+-------+---------+-------+

 cat aa
1,2,ab,cd,ef

curl --location-trusted -u "root:" -T aa  -H 'column_separator:,'  -H "columns:c1,c2,A,B,C,c3=CONCAT(A,B,C)"  http://127.0.0.1:8030/api/test/tbl/_stream_load

MySQL [test]> select * from tbl;
+------+------+--------+
| c1   | c2   | c3     |
+------+------+--------+
|    1 |    2 | abcdef |
+------+------+--------+

**Reply by flyzyh** [reply]
感谢两位大佬回复
我的拼接内容是包含下划线的，就是column_u=A_B_C，拼接再加上下划线，还要用shell的curl，就有点恶心了。
还是感谢大佬

**Reply by liaoxin01** [reply]
带下划线，可以使用CONCAT_WS函数
curl --location-trusted -u "root:" -T aa  -H 'column_separator:,'  -H "columns:c1,c2,A,B,C,c3=CONCAT_WS('_', A,B,C)"  http://127.0.0.1:8030/api/test/tbl/_stream_load

MySQL [test]> select * from test.tbl
+------+------+----------+
| c1   | c2   | c3       |
+------+------+----------+
|    1 |    2 | abcdef   |
|    1 |    2 | ab_cd_ef |
+------+------+----------+

你也可以不用shell的curl，使用各种编程语言都可以，按照这种api去写就行

**Reply by flyzyh** [reply]
谢谢大佬。
之所以考虑用shell，是因为公司对代码有审计，但是对curl的脚本就宽松一些。
我们现在就是用insert操作的，已经绕过了这些问题，代价就是稍有改动特别麻烦。
